File: engine/constraints.js
/**
 * STEP 5 — Constraint Engine
 *
 * Constraints object format (from frontend):
 * {
 *   x: { min?: number, max?: number, even?: boolean, eq?: number }
 * }
 *
 * This module validates/normalizes constraints and provides helpers for pruning.
 */

/** @typedef {{ min?: number, max?: number, even?: boolean, eq?: number }} VarConstraint */
/** @typedef {Record<string, VarConstraint>} Constraints */

function assertInt(n, label) {
  if (!Number.isSafeInteger(n)) throw new Error(`${label} must be a safe integer.`);
}

function normalizeVarConstraint(variable, c) {
  /** @type {VarConstraint} */
  const out = {};

  if (c == null) return out;
  if (typeof c !== "object") throw new Error(`Constraint for ${variable} must be an object.`);

  if (c.min != null) {
    assertInt(c.min, `${variable}.min`);
    out.min = c.min;
  }
  if (c.max != null) {
    assertInt(c.max, `${variable}.max`);
    out.max = c.max;
  }
  if (c.eq != null) {
    assertInt(c.eq, `${variable}.eq`);
    out.eq = c.eq;
  }
  if (c.even != null) {
    out.even = Boolean(c.even);
  }

  // Global rule: non-negative values only.
  if (out.min == null || out.min < 0) out.min = 0;
  if (out.max != null && out.max < 0) {
    // If max < 0 but values must be >= 0, domain becomes empty.
    out.max = -1;
  }
  if (out.eq != null && out.eq < 0) {
    out.eq = -1;
  }

  // If eq is present, it overrides min/max.
  if (out.eq != null) {
    out.min = out.eq;
    out.max = out.eq;
  }

  if (out.min != null && out.max != null && out.min > out.max) {
    throw new Error(`Invalid constraint range for ${variable}: min > max.`);
  }

  if (out.even && out.min != null && out.max != null && out.max >= 0) {
    // Domain can still be empty but we'll let domain calculation handle it.
  }

  return out;
}

/**
 * Normalize full constraints map.
 * @param {Constraints | undefined | null} constraints
 * @returns {Constraints}
 */
function normalizeConstraints(constraints) {
  if (constraints == null) return {};
  if (typeof constraints !== "object") throw new Error("constraints must be an object.");

  /** @type {Constraints} */
  const out = {};
  for (const [variable, c] of Object.entries(constraints)) {
    out[variable] = normalizeVarConstraint(variable, c);
  }
  return out;
}

/**
 * Compute allowed integer domain for a variable, optionally capped by a maximum.
 * @param {string} variable
 * @param {Constraints} constraints
 * @param {number} hardMax A cap computed from remaining/a (>= 0)
 */
function domainFor(variable, constraints, hardMax) {
  const c = constraints[variable] ?? {};

  const min = c.min ?? 0;
  const max = c.max != null ? Math.min(c.max, hardMax) : hardMax;

  if (max < min) return { min: 1, max: 0, step: 1 }; // empty

  const even = Boolean(c.even);
  if (!even) return { min, max, step: 1 };

  // Adjust to even min.
  let adjMin = min;
  if (adjMin % 2 !== 0) adjMin += 1;

  // Adjust max downward to even max.
  let adjMax = max;
  if (adjMax % 2 !== 0) adjMax -= 1;

  if (adjMax < adjMin) return { min: 1, max: 0, step: 2 }; // empty
  return { min: adjMin, max: adjMax, step: 2 };
}

/**
 * Check whether a fully-assigned solution satisfies constraints.
 * @param {Record<string, number>} solution
 * @param {Constraints} constraints
 */
function satisfiesAll(solution, constraints) {
  for (const [variable, c] of Object.entries(constraints)) {
    if (!c) continue;
    const value = solution[variable];
    if (value == null) continue; // variables not in equation may be ignored

    if (c.min != null && value < c.min) return false;
    if (c.max != null && value > c.max) return false;
    if (c.eq != null && value !== c.eq) return false;
    if (c.even && value % 2 !== 0) return false;
  }
  return true;
}

module.exports = {
  normalizeConstraints,
  domainFor,
  satisfiesAll,
};

File: engine/rust/src/constraints.rs
// ============================================================
//  constraints.rs — Port of engine/constraints.js
//
//  Parses & normalizes the variable constraint map, and
//  provides helpers used by the solver during pruning.
// ============================================================

use std::collections::HashMap;
use serde::Deserialize;

/// Per-variable constraint (all fields optional)
#[derive(Debug, Clone, Default, Deserialize)]
pub struct VarConstraint {
    pub min:  Option<i64>,
    pub max:  Option<i64>,
    pub eq:   Option<i64>,
    pub even: Option<bool>,
}

pub type Constraints = HashMap<char, VarConstraint>;

/// Allowed integer domain for a variable during backtracking.
#[derive(Debug, Clone)]
pub struct Domain {
    pub min:  i64,
    pub max:  i64,
    pub step: i64,
}

impl Domain {
    pub fn is_empty(&self) -> bool {
        self.max < self.min
    }
}

/// Normalize a single variable constraint.
///   - Floors min at 0 (non-negative requirement).
///   - `eq` overrides min & max.
///   - Validates min <= max.
pub fn normalize_var_constraint(c: VarConstraint) -> Result<VarConstraint, String> {
    let mut out = VarConstraint::default();

    // Copy fields
    if let Some(min) = c.min { out.min = Some(min); }
    if let Some(max) = c.max { out.max = Some(max); }
    if let Some(eq)  = c.eq  { out.eq  = Some(eq);  }
    if let Some(even) = c.even { out.even = Some(even); }

    // Non-negative floor
    if out.min.unwrap_or(0) < 0 { out.min = Some(0); }
    if let Some(max) = out.max { if max < 0 { out.max = Some(-1); } }
    if let Some(eq)  = out.eq  { if eq  < 0 { out.eq  = Some(-1); } }

    // eq overrides min/max
    if let Some(eq) = out.eq {
        out.min = Some(eq);
        out.max = Some(eq);
    }

    if let (Some(min), Some(max)) = (out.min, out.max) {
        if min > max {
            return Err(format!("Invalid constraint: min ({}) > max ({}).", min, max));
        }
    }

    Ok(out)
}

/// Parse a JSON string `{"x":{"min":1,"max":10}}` → Constraints.
/// Keys must be single characters.
pub fn parse_constraints_json(json: &str) -> Result<Constraints, String> {
    if json.trim().is_empty() || json.trim() == "{}" || json.trim() == "null" {
        return Ok(HashMap::new());
    }

    let raw: HashMap<String, VarConstraint> = serde_json::from_str(json)
        .map_err(|e| format!("Invalid constraints JSON: {}", e))?;

    let mut out = Constraints::new();
    for (key, val) in raw {
        let ch = key.chars().next().ok_or_else(|| "Empty constraint key.".to_string())?;
        out.insert(ch, normalize_var_constraint(val)?);
    }
    Ok(out)
}

/// Compute the integer domain for `variable` given constraints and a hard upper bound
/// derived from the remaining sum (remaining / coeff).
pub fn domain_for(var: char, constraints: &Constraints, hard_max: i64) -> Domain {
    let c = constraints.get(&var);

    let min = c.and_then(|c| c.min).unwrap_or(0);
    let max = {
        let cmax = c.and_then(|c| c.max);
        match cmax {
            Some(cm) => cm.min(hard_max),
            None => hard_max,
        }
    };

    if max < min {
        return Domain { min: 1, max: 0, step: 1 }; // empty
    }

    let even = c.and_then(|c| c.even).unwrap_or(false);
    if !even {
        return Domain { min, max, step: 1 };
    }

    // Adjust to nearest even min
    let adj_min = if min % 2 != 0 { min + 1 } else { min };
    // Adjust to nearest even max
    let adj_max = if max % 2 != 0 { max - 1 } else { max };

    if adj_max < adj_min {
        return Domain { min: 1, max: 0, step: 2 };
    }
    Domain { min: adj_min, max: adj_max, step: 2 }
}

/// Check whether a fully-assigned solution satisfies all constraints.
pub fn satisfies_all(solution: &HashMap<char, i64>, constraints: &Constraints) -> bool {
    for (var, c) in constraints {
        if let Some(&val) = solution.get(var) {
            if let Some(min) = c.min { if val < min { return false; } }
            if let Some(max) = c.max { if val > max { return false; } }
            if let Some(eq)  = c.eq  { if val != eq  { return false; } }
            if c.even.unwrap_or(false) && val % 2 != 0 { return false; }
        }
    }
    true
}
